code/modules/integrated_electronics/subtypes/input.dm

```diff
@@ -201,6 +273,50 @@
 		set_pin_data(IC_OUTPUT, 1, pick(valid_things))
 		push_data()
 
+/obj/item/integrated_circuit/input/advanced_locator
+	complexity = 6
+	name = "advanced locator"
+	desc = "This is needed for certain devices that demand a reference for a target to act upon. This type locates something that is standing in given radius of up to 7 meters."
+	extended_desc = "The first pin requires a ref to a kind of object that you want the locator to acquire. This means that it will give refs to nearby objects that are similar to given sample. If this pin is a string, the locator will search for item by matching desired text in name + description. If more than one valid object is found nearby, it will choose one of them at random. The second pin is a radius."
+	inputs = list("desired type" = IC_PINTYPE_ANY, "radius" = IC_PINTYPE_NUMBER)
+	outputs = list("located ref" = IC_PINTYPE_REF )
+	activators = list("locate" = IC_PINTYPE_PULSE_IN,"found" = IC_PINTYPE_PULSE_OUT,"not found" = IC_PINTYPE_PULSE_OUT)
+	spawn_flags = IC_SPAWN_DEFAULT|IC_SPAWN_RESEARCH
+	power_draw_per_use = 30
+	var/radius = 1
+
+/obj/item/integrated_circuit/input/advanced_locator/on_data_written()
+	var/rad = get_pin_data(IC_INPUT, 2)
+	if(isnum(rad))
+		rad = Clamp(rad, 0, 7)
+		radius = rad
+
+/obj/item/integrated_circuit/input/advanced_locator/do_work()
+	set_pin_data(IC_OUTPUT, 1, null)
+	var/turf/T = get_turf(src)
+	var/list/nearby_things = view(radius, T)
+	var/list/valid_things = list()
+	var/I = get_pin_data(IC_INPUT, 1)
+	if(isweakref(I))
+		var/datum/weakref/WR = I
+		var/atom/A = WR.resolve()
+		var/desired_type = A.type
```

Comment:
Weakref might resolve to `null`, so should check that psure.

Reply:
Fair enough, will fix.